### PR TITLE
Utility for scheduling a task and retrieving its CompletableFuture

### LINKED ISSUE #3698: Utility for scheduling a task and retrieving its `CompletableFuture`
It is currently way too verbose to schedule a task with an `EventLoop` or `ScheduledExecutorService` and getting its `CompletableFuture`:

```java
@Get
CompletableFuture<HttpResponse> delayedResponse(ServiceRequestContext ctx) {
  ScheduledFuture<HttpResponse> sf = ctx.eventLoop.schedule(() -> {
    .. do something ..
    return HttpResponse.of(...);
  }, 5, TimeUnit.SECONDS);
  CompletableFuture<HttpResponse> f = new CompletableFuture<>();
  sf.addListener(
    ...
    .. propagate the result to `f` ..
  );
  return f;
}
```

#3697 proposes an API change that simplifies it fundamentally in 2.0, but we could still add similar functionality by adding a utility class like this:

```java
public final class DelayedTasks {
  public static CompletableFuture<?> schedule(Runnable task, Duration delay, Executor executor) { ... }
  ...
}
```

Hence the simplified version of the above example:

```java
@Get
CompletableFuture<HttpResponse> delayedResponse(ServiceRequestContext ctx) {
  return DelayedTasks.schedule(() -> {
    .. do something ..
    return HttpResponse.of(...)
  }, Duration.ofSeconds(5), ctx.eventLoop());
}
```

### PR DESCRIPTION
**Pull request by TheWeaVer**
Motivation:

- Simplify of the scheduling a task with an `EventLoop` or `ScheduledExecutorService`

Modifications:

- Add two overloaded methods of `HttpResponse.delayed()`

Result:

- Closes #3698

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java [hunk 1]
```diff
@@ -154,6 +155,31 @@ static HttpResponse delayed(HttpResponse response, Duration delay, ScheduledExec
...
+     * delegates to the provided {@link HttpResponse} {@link Supplier},
+     * beginning publishing after {@code delay} has passed from the provided {@link ScheduledExecutorService}.
+     */
+    static HttpResponse delayed(Supplier<? extends HttpResponse> responseSupplier,
+                                Duration delay,
+                                ScheduledExecutorService executor) {
+        requireNonNull(responseSupplier, "responseSupplier");
+        requireNonNull(delay, "delay");
+        requireNonNull(executor, "executor");
+        return delayed(responseSupplier.get(), delay, executor);
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:180]: We need to lazily invoke `responseSupplier.get()` when the `delay` comes like:
```java
executor.schedule(() -> responseSupplier.get().delegate(response), delay.toNanos(), TimeUnit.NANOSECONDS);
```

Comment (TheWeaVer) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponse.java]: I applied it, thank you 😄

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java [hunk 2]
```diff
@@ -154,6 +155,33 @@ static HttpResponse delayed(HttpResponse response, Duration delay, ScheduledExec
...
+    }
+
+    /***
+     * Invokes the specified {@link Supplier} and creates a new HTTP response that
+     * delegates to the provided {@link HttpResponse} {@link Supplier},
+     * beginning publishing after {@code delay} has passed from the provided {@link ScheduledExecutorService}.
+     */
+    static HttpResponse delayed(Supplier<? extends HttpResponse> responseSupplier,
+                                Duration delay,
+                                ScheduledExecutorService executor) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:176]: How about migrating `delayed(HttpResponse response, Duration delay, ScheduledExecutorService executor)` to use this?
For example:
```java
delay(() -> response, delay, executor);
```

**Review by ikhoon — COMMENTED** [review]
Could add some tests for the changes?

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java (hunk 2 above)
Comment (TheWeaVer) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:176]: Thank you, let me apply it and make some test cases 🙂

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java [hunk 3]
```diff
@@ -154,6 +155,33 @@ static HttpResponse delayed(HttpResponse response, Duration delay, ScheduledExec
...
 
+    /***
+     * Invokes the specified {@link Supplier} and creates a new HTTP response that
+     * delegates to the provided {@link HttpResponse} by {@link Supplier},
+     * beginning publishing after {@code delay} has passed from a random {@link ScheduledExecutorService}.
+     */
+    static HttpResponse delayed(Supplier<? extends HttpResponse> responseSupplier, Duration delay) {
+        requireNonNull(responseSupplier, "responseSupplier");
+        requireNonNull(delay, "delay");
+        return delayed(responseSupplier, delay, CommonPools.workerGroup().next());
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:166]: Should we use the current thread-local `RequestContext`'s event loop if possible?

Comment (TheWeaVer) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:166]: Then is it okay to replace `CommonPools.workerGroup().next()` to `RequestContext.current().eventLoop()`? What is the meaning of the if possible?.?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:166]: You may want to see https://github.com/line/armeria/blob/3e2185083fdd7ad45b9b7a71cf5b573cbd8791d2/core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java#L442

Comment (TheWeaVer) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:166]: Fully what I want! Thank you 🙂

File: core/src/test/java/com/linecorp/armeria/common/HttpResponseTest.java [hunk 4]
```diff
@@ -150,4 +154,56 @@ void ofRedirectParamsShouldNotBeNull() {
             null))
             .isInstanceOf(NullPointerException.class).hasMessageContaining("args");
     }
+
+    @Test
+    void delayedHttpResponseWithAggregatedHttpResponseUsingCurrentEventLoopOrCommonPools() {
+        final AggregatedHttpResponse aggregatedHttpResponse = AggregatedHttpResponse.of(HttpStatus.OK);
+        final HttpResponse res = HttpResponse.delayed(aggregatedHttpResponse, Duration.ofSeconds(1));
+        assertThat(res.aggregate().join().status()).isEqualTo(HttpStatus.OK);
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/common/HttpResponseTest.java:162]: Could we check if 1 second has actually passed?
Perhaps use Guava's `StopWatch`?

**Review by ikhoon — APPROVED** [review]
Thanks! @TheWeaVer 
Left a minor nit.

File: core/src/test/java/com/linecorp/armeria/common/HttpResponseTest.java (hunk 4 above)
Comment (TheWeaVer) [reply core/src/test/java/com/linecorp/armeria/common/HttpResponseTest.java:162]: Thank you 😄 Let me apply it !

**Comment by minwoox** [comment]
So we don't need the `DelayedTasks` which is not just for the `HttpResponse` but for any classes? @trustin?

**Review by ikhoon — APPROVED** [review]
Thanks! @TheWeaVer Looking forward to your next! 🚀

**Review by trustin — APPROVED** [review]
:+1:

**Comment by trustin** [comment]
@minwoox Please review.

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java [hunk 5]
```diff
@@ -149,8 +150,35 @@ static HttpResponse delayed(HttpResponse response, Duration delay, ScheduledExec
...
+     */
+    static HttpResponse delayed(Supplier<? extends HttpResponse> responseSupplier, Duration delay) {
+        requireNonNull(responseSupplier, "responseSupplier");
+        requireNonNull(delay, "delay");
+        return delayed(responseSupplier, delay,
+                       RequestContext.mapCurrent(RequestContext::eventLoop,
+                                                 CommonPools.workerGroup()::next));
+    }
+
+    /***
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:169]:
```suggestion
    /**
```

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java [hunk 6]
```diff
@@ -149,8 +150,35 @@ static HttpResponse delayed(HttpResponse response, Duration delay, ScheduledExec
         requireNonNull(response, "response");
         requireNonNull(delay, "delay");
         requireNonNull(executor, "executor");
+        return delayed(() -> response, delay, executor);
+    }
+
+    /***
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:156]:
```suggestion
    /**
```

**Review by minwoox — APPROVED** [review]
Thanks a lot, @TheWeaVer 😄

**Pull request merged**